File: cwe_comparative/CWE-89_sql_injection/vulnerable.ts
// CWE-89: SQL Injection
// TypeScript version - vulnerable with string interpolation

// VULNERABLE: Template literal with user input
function getUserVulnerable(username: string): string {
    // BUG: Direct string interpolation
    const query = `SELECT * FROM users WHERE username = '${username}'`;
    return query;
}

// SAFE: Using parameterized queries (pseudo-code)
function getUserSafe(username: string): { query: string; params: string[] } {
    return {
        query: "SELECT * FROM users WHERE username = ?",
        params: [username]
    };
}

// Demonstration
const maliciousInput = "'; DROP TABLE users; --";

console.log("Vulnerable query:");
console.log(getUserVulnerable(maliciousInput));
// Output: SELECT * FROM users WHERE username = ''; DROP TABLE users; --'

console.log("\nSafe query (parameterized):");
const safe = getUserSafe(maliciousInput);
console.log(safe.query);
console.log("Parameters:", safe.params);

File: cwe_comparative/CWE-89_sql_injection/vulnerable.go
// CWE-89: SQL Injection
// Go version - vulnerable if using string formatting

package main

import (
	"database/sql"
	"fmt"
)

// VULNERABLE: String formatting with user input
func getUserVulnerable(db *sql.DB, username string) {
	// BUG: Direct string interpolation
	query := fmt.Sprintf("SELECT * FROM users WHERE username = '%s'", username)
	rows, _ := db.Query(query)
	defer rows.Close()
}

// SAFE: Using parameterized queries
func getUserSafe(db *sql.DB, username string) {
	// Parameterized query - database driver escapes the input
	rows, _ := db.Query("SELECT * FROM users WHERE username = ?", username)
	defer rows.Close()
}

func main() {
	// Demonstration only - no actual DB connection
	fmt.Println("Vulnerable query:")
	fmt.Printf("SELECT * FROM users WHERE username = '%s'\n", "'; DROP TABLE users; --")

	fmt.Println("\nSafe query (parameterized):")
	fmt.Println("SELECT * FROM users WHERE username = ?")
	fmt.Println("Parameter: '; DROP TABLE users; --")
}
